File: C++/PalindromeLinkedList.cc
// Given a singly linked list, determine if it is a palindrome.

// Follow up:
// Could you do it in O(n) time and O(1) space?

class Solution {
    ListNode* reverseList(ListNode* head, ListNode* next) {
        if (!next) return head;
        else {
            ListNode* temp = next->next;
            next->next = head;
            return reverseList(next, temp);
        }
    }
public:
    bool isPalindrome(ListNode* head) {
        if (!head) return true;
        int cnt = 1;
        ListNode* node = head;
        while (node->next) {
            cnt++;
            node = node->next;
        }
        node = head;
        for(int i = 1; i <= cnt/2; i++) node = node->next;
        ListNode* mid = node;
        ListNode* newHead = reverseList(mid, mid->next);
        mid->next = nullptr;
        while(head != newHead) {
            if (head->val != newHead->val) return false;
            if (head->next == newHead) break; 
            head = head->next;
            newHead = newHead->next;
        }
        return true;
    }
};

File: Java/PalindromeLinkedList.java
// Given a singly linked list, determine if it is a palindrome.

public class PalindromeLinkedList {
    ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode n = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return n;
    }

    public boolean isPalindrome(ListNode head) {
        int len = 0;
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
            if (fast != null) {
                fast = fast.next;
            }
        }
        ListNode rev = reverseList(slow);
        while (head != null && rev != null) {
            if (head.val != rev.val) {
                return false;
            }
            head = head.next;
            rev = rev.next;
        }
        return true;
    }
}
